zebra/zebra_nhg.h

```diff
@@ -24,6 +24,245 @@
 #define __ZEBRA_NHG_H__
 
 #include "zebra/rib.h"
+#include "lib/nexthop_group.h"
+
+#include "zebra/zebra_dplane.h"
+
+/* This struct is used exclusively for dataplane
+ * interaction via a dataplane context.
+ *
+ * It is designed to mimic the netlink nexthop_grp
+ * struct in include/linux/nexthop.h
+ */
+struct nh_grp {
+	uint32_t id;
+	uint8_t weight;
+};
+
+PREDECL_RBTREE_UNIQ(nhg_connected_tree);
+
+
+/*
+ * Hashtables contiaining entries found in `zebra_router`.
+ */
+
+struct nhg_hash_entry {
```

Comment:
we'll have nh_grp, nexthop_grp, nhg_ctx, nhg_hash_entry, and nexthop_group structs - that's a bit confusing. can we keep the netlink-specific things in a netlink-specific place (or eliminate them)?  could we extend the existing nexthop_group struct to include the extra things that have been put into `nhg_hash_entry`? the special "this is not a group" case in `nhg_hash_entry` - couldn't we eliminate that by always having a real nexthop, even if there's just one?

Reply:
> we'll have nh_grp, nexthop_grp, nhg_ctx, nhg_hash_entry, and nexthop_group structs - that's a bit confusing. can we keep the netlink-specific things in a netlink-specific place (or eliminate them)? 
See my other reply for info on why the `nh_grp` array. But I believe the netlink specific stuff is separate. I re-used a similar struct as it yes, but that was intentional because I wanted to have the notion of arrays of IDs/Weights inside the `nhg_ctx` object without having to allocate a nexthop struct for all of them.

> could we extend the existing nexthop_group struct to include the extra things that have been put into `nhg_hash_entry`? 
That is a lot of extra memory to be putting into a lib struct that the other daemons will not be using for a while though. I assume this is why @donaldsharp went with the `wrapper` approach rather than putting it directly into a lib struct.

> the special "this is not a group" case in `nhg_hash_entry` - couldn't we eliminate that by always having a real nexthop, even if there's just one?

I am not sure what special case you are referring to. There is always a real `nexthop` and `nexthop_group` contained in the `nhg_hash_entry`